### ISSUE TITLE
Weird behavior of the grad() function when passed an unusual model

### ISSUE THREAD

**Issue by George3d6** [denizyuret/Knet.jl#307]
I tried adding a bit of weird logic to the propagation function in my model.
Calling the function as such:

`predict(model, X)
`
works just fine,

However, when I try to call the model from within the loss function I pass to `grad` I get a weird error within the actual `predict` function:

`
Cannot `convert` an object of type AutoGrad.Rec{Float64} to an object of type Float32
`

Any idea what this error could be coming from, based on the type I assume it's related to the grad() function but I can't figure out how grad would be modifying the inner logic of my predict function to result in this error.

**Comment by CarloLucibello** [comment]
can you post a minimal example?

**Comment by George3d6** [comment]
The specific function I had was this (used to try and generate music samples):

```
function generate(Model, Z)
function generate(Model, Z=nothing;)
    for i=1:2:length(Model) - 2
        Z = leakyrelu.(Model[i]*Z .+ Model[i+1])
    end
    Z = sigm.(Model[end - 1]*Z .+ Model[end])
    
    if "$(typeof(Z[1,1]))" == "AutoGrad.Rec{Float32}" || "$(typeof(Z[1,1]))" =="AutoGrad.Rec{Float64}"
        #return Z 
    end
    
    samples = Array(zeros(Float32, sample_size, length(Z[1,:]) ))
    
    for i=1:nr_smaples
        vals = [] 
        for sample_nr=0:oscilators:oscilators*(granularity - 1)
            push!(vals, Z[(sample_nr + 1):(sample_nr + oscilators), i])        
        end

        for si=1:ample_size
            oi = 1
            for v in vals[Int64(round(1 + si/ample_size * (granularity -1) ))]
                samples[si, i] += sin(si*(oi^(2.3))/16000)*v
                oi += 1
            end
            samples[si, i] = samples[si, i]/oscilators
        end
    end
    return samples
end
```

The problem is that the "neural network" part has an output of a size much smaller then the actual thing I wanted it to generate (basically I wanted to generate X samples of music using a combination of oscilators).

I'm not sure how I can reduce the code so that it's more understandable, probably something like:

```
function generate(Model, Z=nothing;)
    for i=1:2:length(Model) - 2
        Z = leakyrelu.(Model[i]*Z .+ Model[i+1])
    end
    Z = sigm.(Model[end - 1]*Z .+ Model[end])
    
    if "$(typeof(Z[1,1]))" == "AutoGrad.Rec{Float32}" || "$(typeof(Z[1,1]))" =="AutoGrad.Rec{Float64}"
        #return Z 
    end
    
    samples = Array(zeros(Float32, length(Z[1,:]) * 100, length(Z[:,1]) ))
    generate(Z,samples) # Assigns some values to each element of samples based on the values in Z and the position on the matrix of samples
    samples
end
```

But I think this question should be addressed in the AutoGrad pkg and I don't think it's necessarily something that can be solved using autoGrad, I probably just have to write a manual solution, so feel free to close this.
